### PR TITLE
Agharta/eip145 bitwise shift

### PR DESCRIPTION
**Pull request by dziabko**
Implemented EIP 145 (SAR, SHR, SHL opcodes). Implemented instruction_test.go file for basic tests. Added IsAghartha check.

### CODE REVIEW

File: tests/util.go [hunk 1]
```diff
@@ -167,6 +167,13 @@ type StateTest struct {
 	Post map[string][]stPostState `json:"post"`
 }
 
+type ShiftTest struct {
```
Comment (austinabell) [inline tests/util.go:170]: Not used?

File: tests/init.go [hunk 2]
```diff
@@ -141,10 +141,18 @@ var Forks = map[string]RuleSet{
...
+		HomesteadGasRepriceBlock: big.NewInt(0),
+		DiehardBlock:             big.NewInt(0),
+		AtlantisBlock:            big.NewInt(0),
+		AghartaBlock:             big.NewInt(0),
+	},
 	"ConstantinopleFix": {
 		HomesteadBlock:           big.NewInt(0),
 		HomesteadGasRepriceBlock: big.NewInt(0),
 		DiehardBlock:             big.NewInt(0),
+		ExplosionBlock:           big.NewInt(0),
```
Comment (austinabell) [inline tests/init.go]: Not needed to define ExplosionBlock

**Review by austinabell — CHANGES_REQUESTED** [review]

**Comment by soc1c** [comment]
changed base to `development`, waiting for reapplication of #5 and #7

File: core/vm/gas.go [hunk 3]
```diff
@@ -123,61 +123,65 @@ type req struct {
...
+	GAS:          {0, GasQuickStep, 1},
+	BLOCKHASH:    {1, GasExtStep, 1},
+	BALANCE:      {1, new(big.Int), 1},
+	EXTCODESIZE:  {1, new(big.Int), 1},
+	EXTCODECOPY:  {4, new(big.Int), 0},
+	SLOAD:        {1, big.NewInt(50), 1},
+	SSTORE:       {2, new(big.Int), 0},
+	SHA3:         {2, big.NewInt(30), 1},
+	CREATE:       {3, big.NewInt(32000), 1},
 	RETURNDATASIZE: {0, GasQuickStep, 1},
```
Comment (austinabell) [inline core/vm/gas.go]: fmt if possible

File: core/execution.go [hunk 4]
```diff
@@ -223,7 +224,111 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
...
+	env.Db().SetNonce(caller.Address(), nonce+1)
+
+	// Create a new account on the state
+	snapshot := env.SnapshotDatabase()
+
+	//Create account with address
+	to := env.Db().CreateAccount(*address)
+
+	if env.RuleSet().IsAtlantis(env.BlockNumber()) {
+		env.Db().SetNonce(*address, 1)
```
Comment (austinabell) [inline core/execution.go]: Reintrocuces nonce bug, do starting nonce + 1.

File: core/vm/instructions.go [hunk 5]
```diff
@@ -523,6 +570,37 @@ func opCreate(pc *uint64, env Environment, contract *Contract, memory *Memory, s
 	return nil, nil
 }
 
+func opCreate2(pc *uint64, env Environment, contract *Contract, memory *Memory, stack *stack) ([]byte, error) {
```
Comment (austinabell) [inline core/vm/instructions.go:573]: Why is Create2 being introduced with bitwise shift?

File: core/vm/runtime/env.go [hunk 6]
```diff
@@ -115,3 +115,7 @@ func (self *Env) StaticCall(me vm.ContractRef, addr common.Address, data []byte,
 func (self *Env) Create(caller vm.ContractRef, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error) {
 	return core.Create(self, caller, data, gas, price, value)
 }
+
+func (self *Env) Create2(caller vm.ContractRef, data []byte, gas, price, salt, value *big.Int) ([]byte, common.Address, error) {
```
Comment (austinabell) [inline core/vm/runtime/env.go:119]: same as above

File: core/vm/vm.go [hunk 7]
```diff
@@ -334,6 +335,12 @@ func calculateGasAndSize(gasTable *GasTable, env Environment, contract *Contract
 	case CREATE:
 		newMemSize = calcMemSize(stack.back(1), stack.back(2))
 
+		quadMemGas(mem, newMemSize, gas)
+	case CREATE2:
```
Comment (austinabell) [inline core/vm/vm.go:339]: same as above

File: core/vm_env.go [hunk 8]
```diff
@@ -122,3 +122,7 @@ func (self *VMEnv) StaticCall(me vm.ContractRef, addr common.Address, data []byt
 func (self *VMEnv) Create(me vm.ContractRef, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error) {
 	return Create(self, me, data, gas, price, value)
 }
+
+func (self *VMEnv) Create2(me vm.ContractRef, data []byte, gas, price, salt, value *big.Int) ([]byte, common.Address, error) {
```
Comment (austinabell) [inline core/vm_env.go:126]: same as above

File: crypto/crypto.go [hunk 9]
```diff
@@ -62,6 +62,12 @@ func CreateAddress(b common.Address, nonce uint64) common.Address {
 	return common.BytesToAddress(Keccak256(data)[12:])
 }
 
+// CreateAddress2 creates an ethereum address given the address bytes, initial
+// contract code hash and a salt.
+func CreateAddress2(b common.Address, salt []byte, inithash []byte) common.Address {
```
Comment (austinabell) [inline crypto/crypto.go:67]: same as above

File: tests/state_test.go [hunk 10]
```diff
@@ -689,10 +689,40 @@ func TestAllETH(t *testing.T) {
 	skipTests["RevertPrecompiledTouch_storage.json/ConstantinopleFix/0"] = "Bug in Test"
 	skipTests["RevertPrecompiledTouch_storage.json/ConstantinopleFix/3"] = "Bug in Test"
 
+	// EIP 684 Implementations
```
Comment (austinabell) [inline tests/state_test.go]: The following test skips should not be reintroduced or added (All of them)

File: tests/state_test.go [hunk 11]
```diff
@@ -710,8 +740,8 @@ func TestAllETH(t *testing.T) {
 
 func runETHTests(t *testing.T, fileNames []string, skipTests map[string]string) {
 	unsupportedForkConfigs := map[string]bool{
-		"Constantinople":               true,
-		"ConstantinopleFix":            true,
+		"Constantinople": true,
+		// "ConstantinopleFix":            true,
```
Comment (austinabell) [inline tests/state_test.go]: Remove the commented out line if you don't plan to add it back after this

File: tests/util.go [hunk 12]
```diff
@@ -397,6 +409,18 @@ func (self *Env) Create(caller vm.ContractRef, data []byte, gas, price, value *b
 	}
 }
 
+func (self *Env) Create2(caller vm.ContractRef, data []byte, gas, price, salt, value *big.Int) ([]byte, common.Address, error) {
```
Comment (austinabell) [inline tests/util.go:412]: Remove this in the context of this PR

File: core/execution.go [hunk 13]
```diff
@@ -37,6 +37,7 @@ var (
 	maxCodeSize            = 24576
 	errMaxCodeSizeExceeded = fmt.Errorf("Max Code Size exceeded (%d)", maxCodeSize)
 
+	errCodeStoreOutOfGas        = errors.New("contract creation code storage out of gas")
```
Comment (austinabell) [inline core/execution.go]: Not relevant to your changes?

File: cmd/evm/main.go [hunk 14]
```diff
@@ -300,3 +304,7 @@ func (self *VMEnv) StaticCall(caller vm.ContractRef, addr common.Address, data [
 func (self *VMEnv) Create(caller vm.ContractRef, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error) {
 	return core.Create(self, caller, data, gas, price, value)
 }
+
+func (self *VMEnv) Create2(caller vm.ContractRef, data []byte, gas, price, salt, value *big.Int) ([]byte, common.Address, error) {
```
Comment (austinabell) [inline cmd/evm/main.go:308]: This is bitwise shift PR?

File: core/execution.go [hunk 15]
```diff
@@ -223,7 +224,111 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
 
 // Create creates a new contract with the given code
 func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPrice, value *big.Int) (ret []byte, address common.Address, err error) {
-	// Depth check execution. Fail if we're trying to execute above the limit.
+	nonce := env.Db().GetNonce(caller.Address())
```
Comment (austinabell) [inline core/execution.go:226]: Should not be changing the logic of Create, the previous changes were done to fix bugs within a simulated environment

File: core/execution.go [hunk 16]
```diff
@@ -223,7 +224,111 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
...
+	// Assign err if contract code size exceeds the max while the err is still empty.
+	if maxCodeSizeExceeded && err == nil {
+		err = errMaxCodeSizeExceeded
+	}
+
+	return ret, *address, err
+
+}
+
+func exec(env vm.Environment, caller vm.ContractRef, codeAddr *common.Address, codeHash common.Hash, input, code []byte, gas, gasPrice, value *big.Int, readOnly bool) (ret []byte, addr common.Address, err error) {
```
Comment (austinabell) [inline core/execution.go]: this isn't being used from what I can see?

**Review by austinabell — CHANGES_REQUESTED** [review]
Please be careful when and where you rebase and changes that are introduced

**Review by GregTheGreek — DISMISSED** [review]

**Comment by austinabell** [comment]
This PR isn't ready, requires CREATE2 in first and be rebased for tests/ agharta config

**Pull request closed without merging**